README.md
## csg.js

## Constructive Solid Geometry (CSG) Library

 Solid modelling library (2d & 3d)

## Overview

Constructive Solid Geometry (CSG) is a modelling technique that uses Boolean operations like union and intersection to combine 3D solids. This library implements CSG operations on meshes elegantly and concisely using BSP trees, and is meant to serve as an easily understandable implementation of the algorithm.

> The developers of the main project (JSCAD) are busy creating the second version of the API. 
> Because the first version is needed for my own purposes. 
> It was decided to continue developing the first version.

## Table of Contents

- [Installation](#installation)
- [Usage](#usage)
- [API](#api)
- [Contribute](#contribute)
- [License](#license)

## Installation

```
npm install sibvrv/libcsg
```

## Usage

- as Node module :

```
const csg = require('libcsg')
```

## API

The API documentation can be found [here](./docs/api.md).

Also see the [OpenJsCad User Guide](https://en.wikibooks.org/wiki/OpenJSCAD_User_Guide).

## Contribute

This library is part of the JSCAD Organization, and is maintained by a group of volunteers. We welcome and encourage anyone to pitch in but please take a moment to read the following guidelines.

* If you want to submit a bug report please make sure to follow the [Reporting Issues](https://github.com/jscad/csg.js/wiki/Reporting-Issues) guide. Bug reports are accepted as [Issues](https://github.com/sibvrv/libcsg/issues/) via GitHub.

* If you want to submit a change or a patch, please see the [Contributing guidelines](https://github.com/jscad/csg.js/blob/master/CONTRIBUTING.md). New contributions are accepted as [Pull Requests](https://github.com/sibvrv/libcsg/pulls/) via GithHub.

* We only accept bug reports and pull requests on **GitHub**.

* If you have a question about how to use libCSG. You might find the answer in the [OpenJSCAD.org User Guide](https://github.com/Spiritdude/OpenJSCAD.org/wiki/User-Guide).

## Copyrights

Some copyrights apply. 
* Copyright (c) 2012 Joost Nieuwenhuijse (joost@newhouse.nl), under the MIT license. 
* Copyright (c) 2011 Evan Wallace (http://madebyevan.com/)

## License

[The MIT License (MIT)](https://github.com/sibvrv/libcsg/blob/master/LICENSE)

(unless specified otherwise)


CHANGELOG.md
# [0.8.0](https://github.com/sibvrv/libcsg/compare/v0.7.0...v0.8.0) (2020-05-03)


### Bug Fixes

* **center:** fix center function z axis for non-uniform geometries ([#129](https://github.com/sibvrv/libcsg/issues/129)) ([0546043](https://github.com/sibvrv/libcsg/commit/054604309613a40b500f040169b8cd8ed9413343))



<a name="0.7.0"></a>
# [0.7.0](https://github.com/jscad/csg.js/compare/v0.6.0...v0.7.0) (2018-09-01)


### Features

* **path:** Add getTurn()  method to determine the general turn direction of a path. ([#116](https://github.com/jscad/csg.js/issues/116)) ([4cd83fa](https://github.com/jscad/csg.js/commit/4cd83fa))



<a name="0.6.0"></a>
# [0.6.0](https://github.com/jscad/csg.js/compare/v0.5.4...v0.6.0) (2018-06-21)


### Bug Fixes

* **2d shape factories:** fix fromNestedPointsArray([#103](https://github.com/jscad/csg.js/issues/103)) ([729d459](https://github.com/jscad/csg.js/commit/729d459))
* **polygon:** params custom paths ([#104](https://github.com/jscad/csg.js/issues/104)) ([7d93db8](https://github.com/jscad/csg.js/commit/7d93db8))


### Features

* **2d polygons:** add support for polygons (2D) with holes([#101](https://github.com/jscad/csg.js/issues/101)) ([22f8f80](https://github.com/jscad/csg.js/commit/22f8f80))
* **vector_text/char:** big overhaul & additions to the vector_text/ char system ([#107](https://github.com/jscad/csg.js/issues/107)) ([eadcd58](https://github.com/jscad/csg.js/commit/eadcd58))



<a name="0.5.4"></a>
## [0.5.4](https://github.com/jscad/csg.js/compare/v0.5.3...v0.5.4) (2018-03-16)


### Bug Fixes

* **center:** Correcting center() functionality ([#97](https://github.com/jscad/csg.js/issues/97)) ([52e792d](https://github.com/jscad/csg.js/commit/52e792d))



<a name="0.5.3"></a>
## [0.5.3](https://github.com/jscad/csg.js/compare/v0.5.2...v0.5.3) (2018-02-13)



<a name="0.5.2"></a>
## [0.5.2](https://github.com/jscad/csg.js/compare/v0.5.1...v0.5.2) (2018-01-29)


### Bug Fixes

* **rotate:** fixing angle-axis style rotation ([#90](https://github.com/jscad/csg.js/issues/90)) ([6c28c1c](https://github.com/jscad/csg.js/commit/6c28c1c))



<a name="0.5.1"></a>
## [0.5.1](https://github.com/jscad/csg.js/compare/v0.5.0...v0.5.1) (2018-01-29)


### Bug Fixes

* **CAG:** added back missing CAG.fromCompactBinary ([#91](https://github.com/jscad/csg.js/issues/91)) ([57e8f5d](https://github.com/jscad/csg.js/commit/57e8f5d))



<a name="0.5.0"></a>
# [0.5.0](https://github.com/jscad/csg.js/compare/v0.4.1...v0.5.0) (2018-01-19)



<a name="0.4.1"></a>
## [0.4.1](https://github.com/jscad/csg.js/compare/v0.4.0...v0.4.1) (2018-01-19)



<a name="0.4.0"></a>
# [0.4.0](https://github.com/jscad/csg.js/compare/v0.3.8...v0.4.0) (2018-01-05)


### Bug Fixes

* **docstring:** fixed a few bad docstrings which prevented docs from being generated ([#76](https://github.com/jscad/csg.js/issues/76)) ([441fd18](https://github.com/jscad/csg.js/commit/441fd18))



<a name="0.3.8"></a>
## [0.3.8](https://github.com/jscad/csg.js/compare/v0.3.7...v0.3.8) (2018-01-04)


### Bug Fixes

* **shared object:**  support shared objects without color ([#74](https://github.com/jscad/csg.js/issues/74)) ([f90ad50](https://github.com/jscad/csg.js/commit/f90ad50))



<a name="0.3.7"></a>
## [0.3.7](https://github.com/jscad/csg.js/compare/v0.3.6...v0.3.7) (2017-11-13)



<a name="0.3.6"></a>
## [0.3.6](https://github.com/jscad/csg.js/compare/v0.3.5...v0.3.6) (2017-11-03)



<a name="0.3.5"></a>
## [0.3.5](https://github.com/jscad/csg.js/compare/v0.3.4...v0.3.5) (2017-11-03)


### Bug Fixes

* **fixTJunctions:** fixes issues with fixTJunctions ([#63](https://github.com/jscad/csg.js/issues/63)) ([78c5102](https://github.com/jscad/csg.js/commit/78c5102))



<a name="0.3.4"></a>
## [0.3.4](https://github.com/jscad/csg.js/compare/v0.3.3...v0.3.4) (2017-11-01)


### Bug Fixes

* **ConnectorsList:** add back missing ConnectorsList ([#62](https://github.com/jscad/csg.js/issues/62)) ([cfc1c7e](https://github.com/jscad/csg.js/commit/cfc1c7e))



<a name="0.3.3"></a>
## [0.3.3](https://github.com/jscad/csg.js/compare/v0.3.2...v0.3.3) (2017-11-01)



<a name="0.3.2"></a>
## [0.3.2](https://github.com/jscad/csg.js/compare/v0.3.1...v0.3.2) (2017-10-28)


### Bug Fixes

* **path2D:** added missing innerToCAG method to path2D proto & added tests & docs ([#52](https://github.com/jscad/csg.js/issues/52)) ([4a5e37e](https://github.com/jscad/csg.js/commit/4a5e37e))



<a name="0.3.1"></a>
## [0.3.1](https://github.com/jscad/csg.js/compare/v0.3.0...v0.3.1) (2017-06-11)



<a name="0.3.0"></a>
# [0.3.0](https://github.com/jscad/csg.js/compare/v0.2.4...v0.3.0) (2017-05-30)


### Features

* **CAG.toPoints:** CAG Enhancement for toPoints() ([d023243](https://github.com/jscad/csg.js/commit/d023243))



<a name="0.2.4"></a>
## [0.2.4](https://github.com/jscad/csg.js/compare/v0.2.3...v0.2.4) (2017-05-20)


### Bug Fixes

* **CAG:** reverted back CAG to default to canonicalized = false ([68a0558](https://github.com/jscad/csg.js/commit/68a0558))



<a name="0.2.3"></a>
## [0.2.3](https://github.com/jscad/csg.js/compare/v0.2.2...v0.2.3) (2017-05-19)



<a name="0.2.2"></a>
## [0.2.2](https://github.com/jscad/csg.js/compare/v0.2.1...v0.2.2) (2017-05-19)



<a name="0.2.1"></a>
## [0.2.1](https://github.com/jscad/csg.js/compare/v0.2.0...v0.2.1) (2017-04-27)



<a name="0.2.0"></a>
# [0.2.0](https://github.com/jscad/csg.js/compare/v0.1.4...v0.2.0) (2017-04-27)


### Bug Fixes

* **asserts:** fixed issue with asserts in latest ava ([1210f66](https://github.com/jscad/csg.js/commit/1210f66))



<a name="0.1.4"></a>
## [0.1.4](https://github.com/jscad/csg.js/compare/v0.1.3...v0.1.4) (2017-01-27)


### Bug Fixes

* **README:** more attempts at fixing README on npm ... ([a5ae096](https://github.com/jscad/csg.js/commit/a5ae096))



<a name="0.1.3"></a>
## [0.1.3](https://github.com/jscad/csg.js/compare/v0.1.2...v0.1.3) (2017-01-27)



<a name="0.1.2"></a>

src/api/text.js
const defaultFont = require('../fonts/single-line/hershey/simplex.js');

const defaultsVectorParams = {
  xOffset: 0,
  yOffset: 0,
  input: '?',
  align: 'left',
  font: defaultFont,
  height: 14, // == old vector_xxx simplex font height
  lineSpacing: 2.142857142857143, // == 30/14 == old vector_xxx ratio
  letterSpacing: 1,
  extrudeOffset: 0
};

// vectorsXXX parameters handler
function vectorParams(options, input) {
  if (!input && typeof options === 'string') {
    options = {input: options};
  }
  options = options || {};
  let params = Object.assign({}, defaultsVectorParams, options);
  params.input = input || params.input;
  return params;
}

// translate text line
function translateLine(options, line) {
  const {x, y} = Object.assign({x: 0, y: 0}, options || {});
  let segments = line.segments;
  let segment = null;
  let point = null;
  for (let i = 0, il = segments.length; i < il; i++) {
    segment = segments[i];
    for (let j = 0, jl = segment.length; j < jl; j++) {
      point = segment[j];
      segment[j] = [point[0] + x, point[1] + y];
    }
  }
  return line;
}

/** Represents a character as segments
 * @typedef {Object} VectorCharObject
 * @property {Float} width - character width
 * @property {Float} height - character height (uppercase)
 * @property {Array} segments - character segments [[[x, y], ...], ...]
 */

/** Construct a {@link VectorCharObject} from a ascii character whose code is between 31 and 127,
 * if the character is not supported it is replaced by a question mark.
 * @param {Object|String} [options] - options for construction or ascii character
 * @param {Float} [options.xOffset=0] - x offset
 * @param {Float} [options.yOffset=0] - y offset
 * @param {Float} [options.height=21] - font size (uppercase height)
 * @param {Float} [options.extrudeOffset=0] - width of the extrusion that will be applied (manually) after the creation of the character
 * @param {String} [options.input='?'] - ascii character (ignored/overwrited if provided as seconds parameter)
 * @param {String} [char='?'] - ascii character
 * @returns {VectorCharObject}
 *
 * @example
 * let vectorCharObject = vectorChar()
 * or
 * let vectorCharObject = vectorChar('A')
 * or
 * let vectorCharObject = vectorChar({ xOffset: 57 }, 'C')
 * or
 * let vectorCharObject = vectorChar({ xOffset: 78, input: '!' })
 */
function vectorChar(options, char) {
  let {
    xOffset, yOffset, input, font, height, extrudeOffset
  } = vectorParams(options, char);
  let code = input.charCodeAt(0);
  if (!code || !font[code]) {
    code = 63; // 63 => ?
  }
  let glyph = [].concat(font[code]);
  let ratio = (height - extrudeOffset) / font.height;
  let extrudeYOffset = (extrudeOffset / 2);
  let width = glyph.shift() * ratio;
  let segments = [];
  let polyline = [];
  for (let i = 0, il = glyph.length; i < il; i += 2) {
    gx = ratio * glyph[i] + xOffset;
    gy = ratio * glyph[i + 1] + yOffset + extrudeYOffset;
    if (glyph[i] !== undefined) {
      polyline.push([gx, gy]);
      continue;
    }
    segments.push(polyline);
    polyline = [];
    i--;
  }
  if (polyline.length) {
    segments.push(polyline);
  }
  return {width, height, segments};
}

/** Construct an array of character segments from a ascii string whose characters code is between 31 and 127,
 * if one character is not supported it is replaced by a question mark.
 * @param {Object|String} [options] - options for construction or ascii string
 * @param {Float} [options.xOffset=0] - x offset
 * @param {Float} [options.yOffset=0] - y offset
 * @param {Float} [options.height=21] - font size (uppercase height)
 * @param {Float} [options.lineSpacing=1.4] - line spacing expressed as a percentage of font size
 * @param {Float} [options.letterSpacing=1] - extra letter spacing expressed as a percentage of font size
 * @param {String} [options.align='left'] - multi-line text alignement: left, center or right
 * @param {Float} [options.extrudeOffset=0] - width of the extrusion that will be applied (manually) after the creation of the character
 * @param {String} [options.input='?'] - ascii string (ignored/overwrited if provided as seconds parameter)
 * @param {String} [text='?'] - ascii string
 * @returns {Array} characters segments [[[x, y], ...], ...]
 *
 * @example
 * let textSegments = vectorText()
 * or
 * let textSegments = vectorText('OpenJSCAD')
 * or
 * let textSegments = vectorText({ yOffset: -50 }, 'OpenJSCAD')
 * or
 * let textSegments = vectorText({ yOffset: -80, input: 'OpenJSCAD' })
 */
function vectorText(options, text) {
  let {
    xOffset, yOffset, input, font, height, align, extrudeOffset, lineSpacing, letterSpacing
  } = vectorParams(options, text);
  let [x, y] = [xOffset, yOffset];
  let [i, il, char, vect, width, diff] = [];
  let line = {width: 0, segments: []};
  let lines = [];
  let output = [];
  let maxWidth = 0;
  let lineStart = x;
  const pushLine = () => {
    lines.push(line);
    maxWidth = Math.max(maxWidth, line.width);
    line = {width: 0, segments: []};
  };
  for (i = 0, il = input.length; i < il; i++) {
    char = input[i];
    vect = vectorChar({xOffset: x, yOffset: y, font, height, extrudeOffset}, char);
    if (char === '\n') {
      x = lineStart;
      y -= vect.height * lineSpacing;
      pushLine();
      continue;
    }
    width = vect.width * letterSpacing;
    line.width += width;
    x += width;
    if (char !== ' ') {
      line.segments = line.segments.concat(vect.segments);
    }
  }
  if (line.segments.length) {
    pushLine();
  }
  for (i = 0, il = lines.length; i < il; i++) {
    line = lines[i];
    if (maxWidth > line.width) {
      diff = maxWidth - line.width;
      if (align === 'right') {
        line = translateLine({x: diff}, line);
      } else if (align === 'center') {
        line = translateLine({x: diff / 2}, line);
      }
    }
    output = output.concat(line.segments);
  }
  return output;
}

/** Construct a {@link VectorCharObject} from a ascii character whose code is between 31 and 127,
 * if the character is not supported it is replaced by a question mark.
 * @param {Float} x - x offset
 * @param {Float} y - y offset
 * @param {String} char - ascii character
 * @returns {VectorCharObject}
 * @deprecated >= v2

 * @example
 * let vectorCharObject = vector_char(36, 0, 'B')
 */
function vector_char(x, y, char) {
  return vectorChar({xOffset: x, yOffset: y}, char);
}

/** Construct an array of character segments from a ascii string whose characters code is between 31 and 127,
 * if one character is not supported it is replaced by a question mark.
 * @param {Float} x - x offset
 * @param {Float} y - y offset
 * @param {String} text - ascii string
 * @returns {Array} characters segments [[[x, y], ...], ...]
 * @deprecated >= v2
 *
 * @example
 * let textSegments = vector_text(0, -20, 'OpenJSCAD')
 */
function vector_text(x, y, text) {
  return vectorText({xOffset: x, yOffset: y}, text);
}

module.exports = {
  vector_char,
  vector_text,
  vectorChar,
  vectorText
};
